packages/flutter/lib/src/widgets/editable_text.dart

```diff
@@ -242,24 +242,45 @@ class TextEditingController extends ValueNotifier<TextEditingValue> {
   /// ** See code in examples/api/lib/widgets/editable_text/text_editing_controller.1.dart **
   /// {@end-tool}
   TextEditingController({String? text})
-    : super(text == null ? TextEditingValue.empty : TextEditingValue(text: text));
+    : _originalText = text ?? '',
+      super(
+        text == null ? TextEditingValue.empty : TextEditingValue(text: text),
+      );
 
   /// Creates a controller for an editable text field from an initial [TextEditingValue].
   ///
   /// This constructor treats a null [value] argument as if it were
   /// [TextEditingValue.empty].
   TextEditingController.fromValue(TextEditingValue? value)
     : assert(
-        value == null || !value.composing.isValid || value.isComposingRangeValid,
+        value == null ||
+            !value.composing.isValid ||
+            value.isComposingRangeValid,
         'New TextEditingValue $value has an invalid non-empty composing range '
         '${value.composing}. It is recommended to use a valid composing range, '
         'even for readonly text fields.',
       ),
       super(value ?? TextEditingValue.empty);
 
+
+  /// The original text value when this controller was created.
+  ///
+  /// This field is used to track if the text has been modified from its original
+  /// value. It remains unchanged when:
+  /// - [clear] is called
+  /// - [value] is updated directly
+  /// - [text] is updated
```

Comment:
This is somewhat opinionated, pushing me towards thinking that users should implement this themselves.